Mage/src/main/java/mage/MageObjectImpl.java

```diff
@@ -311,4 +312,23 @@ public void removePTCDA() {
     public String toString() {
         return getIdName() + " (" + super.getClass().getSuperclass().getSimpleName() + " -> " + this.getClass().getSimpleName() + ")";
     }
+
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+
+        if (o == null || this.getClass() != o.getClass()) {
+            return false;
+        }
+
+        return Objects.equals(this.getId(), this.getClass().cast(o).getId());
+    }
+
+    @Override
+    public int hashCode() {
+        return 89 * 7 + Objects.hashCode(this.getId());
+    }
```

Comment:
This would currently be used by `PermanentImpl`, `PermanentToken`, `TokenImpl`, `Cardimpl`, and probably more.

I am not sure if this needs to be override by those.